From 96bdf51e88501598feae7320a9dc4085beddd184 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Mon, 28 Jul 2025 14:23:31 +0800
Subject: [PATCH 223/236] mtk: hostapd: add checks for csa/cca/attlm/reconfig
 flag before disabling mld

Add checks for CSA, CCA, ATTLM, and link reconfig flag before disabling mld
Otherwise, the disabled bss might failed to receive finish event and
update its data.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 src/ap/hostapd.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 67fd1c2be..e3eab9066 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -4073,6 +4073,19 @@ int hostapd_disable_mld(struct hostapd_data *hapd)
 		return -1;
 	}
 
+	if (mld->link_reconf_in_progress || mld->removed_links) {
+		wpa_printf(MSG_ERROR, "Cannot disable MLD during link reconfig");
+		return -1;
+	}
+
+	for_each_mld_link(h, hapd) {
+		if (h->csa_in_progress || h->cca_in_progress ||
+		    hostapd_is_attlm_active(h)) {
+			wpa_printf(MSG_ERROR, "Cannot disable MLD during CSA/CCA/ATTLM");
+			return -1;
+		}
+	}
+
 	for_each_mld_link(h, hapd)
 		hostapd_disable_bss(h);
 
-- 
2.45.2

